nifi-nar-bundles/nifi-site-to-site-reporting-bundle/nifi-site-to-site-reporting-task/src/main/java/org/apache/nifi/reporting/SiteToSiteProvenanceReportingTask.java

```diff
@@ -0,0 +1,354 @@
+            return;
+        }
+
+        List<ProvenanceEventRecord> events;
+        try {
+            events = context.getEventAccess().getProvenanceEvents(firstEventId, context.getProperty(BATCH_SIZE).asInteger());
+        } catch (final IOException ioe) {
+            getLogger().error("Failed to retrieve Provenance Events from repository due to: " + ioe.getMessage(), ioe);
+            return;
+        }
+
+        if (events == null || events.isEmpty()) {
+            getLogger().debug("No events to send due to 'events' being null or empty.");
+            return;
+        }
+
+        final String nifiUrl = context.getProperty(INSTANCE_URL).evaluateAttributeExpressions().getValue();
+        URL url;
+        try {
+            url = new URL(nifiUrl);
+        } catch (final MalformedURLException e1) {
+            // already validated
+            throw new AssertionError();
+        }
+
+        final String hostname = url.getHost();
+        final String platform = context.getProperty(PLATFORM).evaluateAttributeExpressions().getValue();
+
+        final Map<String, ?> config = Collections.emptyMap();
+        final JsonBuilderFactory factory = Json.createBuilderFactory(config);
+        final JsonObjectBuilder builder = factory.createObjectBuilder();
+
+        while (events != null && !events.isEmpty()) {
+            final long start = System.nanoTime();
+
+            // Create a JSON array of all the events in the current batch
+            final JsonArrayBuilder arrayBuilder = factory.createArrayBuilder();
+            for (final ProvenanceEventRecord event : events) {
+                arrayBuilder.add(serialize(factory, builder, event, getComponentName(procGroupStatus, event), hostname, url, rootGroupName, platform));
+            }
+            final JsonArray jsonArray = arrayBuilder.build();
+
+            // Send the JSON document for the current batch
+            try {
+                final Transaction transaction = getClient().createTransaction(TransferDirection.SEND);
+                if (transaction == null) {
+                    getLogger().debug("All destination nodes are penalized; will attempt to send data later");
+                    return;
+                }
+
+                final Map<String, String> attributes = new HashMap<>();
+                final String transactionId = UUID.randomUUID().toString();
+                attributes.put("reporting.task.transaction.id", transactionId);
+
+                final byte[] data = jsonArray.toString().getBytes(StandardCharsets.UTF_8);
+                transaction.send(data, attributes);
+                transaction.confirm();
+                transaction.complete();
+
+                final long transferMillis = TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - start);
+                getLogger().info("Successfully sent {} Provenance Events to destination in {} ms; Transaction ID = {}; First Event ID = {}",
```

Comment:
I'm ok with leaving this at INFO - from the point of view of the reporting task, it is an INFO-level event.